File: Leetcode/tree/94_Binary Tree Inorder Traversal/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        self.dfs(res, root)
        return res

    def dfs(self, res, root):
        if not root: return None
        self.dfs(res, root.left)
        res.append(root.val)
        self.dfs(res, root.right)

File: Leetcode/tree/94_Binary Tree Inorder Traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    /*/// recursive
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        dfs(res, root);
        return res;
    }
        private void dfs(List<Integer> res, TreeNode root){
            if(root==null) return;
            dfs(res, root.left);
            res.add(root.val);
            dfs(res, root.right);
        }/*/
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        Stack<TreeNode> stk = new Stack<>();
        TreeNode temp = root;
        while(temp!=null || !stk.isEmpty()){
            while(temp!=null){
                stk.push(temp);
                temp = temp.left;
            }
            temp = stk.pop();
            res.add(temp.val);
            temp = temp.right;
        }
        return res;
    }
}
